src/compile/data/parse.ts

```diff
@@ -48,56 +49,81 @@ function parseRoot(model: Model, sources: Dict<SourceNode>): DataFlowNode {
 /**
  * Parses a transforms array into a chain of connected dataflow nodes.
  */
-export function parseTransformArray(parent: DataFlowNode, model: Model): DataFlowNode {
+export function parseTransformArray(parent: DataFlowNode, model: Model, ancestorParse: AncestorParse): DataFlowNode {
   let lookupCounter = 0;
 
   model.transforms.forEach(t => {
     if (isCalculate(t)) {
       parent = new CalculateNode(parent, t);
+      ancestorParse.set(t.as, 'derived', false);
     } else if (isFilter(t)) {
-      // Automatically add a parse node for filters with filter objects
       const parse = {};
-      const filter = t.filter;
-      let val: string | number | boolean | DateTime = null;
-      // For EqualFilter, just use the equal property.
-      // For RangeFilter and OneOfFilter, all array members should have
-      // the same type, so we only use the first one.
-      if (isFieldEqualPredicate(filter)) {
-        val = filter.equal;
-      } else if (isFieldRangePredicate(filter)) {
-        val = filter.range[0];
-      } else if (isFieldOneOfPredicate(filter)) {
-        val = (filter.oneOf || filter['in'])[0];
-      } // else -- for filter expression, we can't infer anything
-      if (val) {
-        if (isDateTime(val)) {
-          parse[filter['field']] = 'date';
-        } else if (isNumber(val)) {
-          parse[filter['field']] = 'number';
-        } else if (isString(val)) {
-          parse[filter['field']] = 'string';
+      forEachLeaf(t.filter, filter => {
+        if (isFieldPredicate(filter)) {
+          // Automatically add a parse node for filters with filter objects
+          let val: string | number | boolean | DateTime = null;
+
+          // For EqualFilter, just use the equal property.
+          // For RangeFilter and OneOfFilter, all array members should have
+          // the same type, so we only use the first one.
+          if (isFieldEqualPredicate(filter)) {
+            val = filter.equal;
+          } else if (isFieldRangePredicate(filter)) {
+            val = filter.range[0];
+          } else if (isFieldOneOfPredicate(filter)) {
+            val = (filter.oneOf || filter['in'])[0];
+          } // else -- for filter expression, we can't infer anything
+          if (val) {
+            if (isDateTime(val)) {
+              parse[filter.field] = 'date';
+            } else if (isNumber(val)) {
+              parse[filter.field] = 'number';
+            } else if (isString(val)) {
+              parse[filter.field] = 'string';
+            }
+          }
+
+          if (filter.timeUnit) {
+            parse[filter.field] = 'date';
+          }
         }
-      }
+      });
 
-      if (keys(parse).length > 0) {
-        parent = new ParseNode(parent, parse);
+      if (keys(parse).length) {
+        parent = ParseNode.makeWithAncestors(parent, {}, parse, ancestorParse) || parent;
```

Comment:
I find the name `makeImplicit` a bit misleading since this parse from transform is "implicit" too, but doesn't call that `makeImplicit` method.  

May be rename `makeImplicit` to be`makeImplicitFromEncoding` and extract the logic above here to be `makeImplicitFromFilterTransform` (or just FromTransform).   This way you can make `makeWithAncestor` a `private` method too. 


One question that I have is don't we have to do the same thing with TimeUnit and Bin, in addition to Filter?